src/support/allocators/node_allocator.h

```diff
@@ -0,0 +1,399 @@
+
+    /**
+     * Deallocates all allocated blocks.
+     */
+    ~MemoryResource() noexcept
+    {
+        Clear();
+    }
+
+    /**
+     * Clears all allocated blocks.
+     *
+     * Memory is freed regardless if its still in use, so use with care! E.g. only after a call to clear() to the underlying container.
+     */
+    void Clear()
+    {
+        for (auto* block : m_allocated_blocks) {
+            ::operator delete(block);
+        }
+        m_free_chunks = nullptr;
+        m_allocated_blocks.clear();
+        m_chunk_size_bytes = 0U;
+        m_untouched_memory_iterator = nullptr;
+        m_untouched_memory_end = nullptr;
+    }
+
+    /**
+     * Allocates memory for n times T. Only when n==1 the memory blocks are used to give out memory. The first call with n==1
+     * decides the chunk size.
+     *
+     * @tparam T Object to allocate memory for.
+     * @param n Number of objects to allocate for
+     */
+    template <typename T>
+    [[nodiscard]] T* Allocate(size_t n)
+    {
+        // assign to a constexpr variable to force constexpr evaluation
+        static constexpr auto required_chunk_size = CalcRequiredChunkSizeBytes<T>();
+
+        // only use pool when a single element is allocated
+        if (m_chunk_size_bytes == 0 && n == 1) {
+            // first call to Allocate with n==1 determines chunk size
+            m_chunk_size_bytes = required_chunk_size;
+        }
+
+        if (m_chunk_size_bytes != required_chunk_size || n != 1) {
+            // pool is not used so forward to operator new.
+            return static_cast<T*>(::operator new(n * sizeof(T)));
+        }
+
+        // chunk size is correct, so we can actually use the pool's block data
+
+        if (m_free_chunks) {
+            // we've already got data in the free list, unlink one element
+            auto old_head = m_free_chunks;
+            m_free_chunks = static_cast<ChunkNode*>(m_free_chunks)->next;
+            return static_cast<T*>(old_head);
+        }
+
+        // free list is empty: get one chunk from allocated block memory.
+        // It makes sense to not create the fully linked list of an allocated block up front, for several reasons
+        // On the one hand, the latency is higher when we need to iterate and update pointers for the whole block at once.
+        // More importantly, most systems lazily allocate data. So when we allocate a big block of memory the memory for a page
+        // is only actually made available to the program when it is first touched. So when we allocate a big block and only use
+        // very little memory from it, the total memory usage is lower than what has been malloc'ed.
+        if (m_untouched_memory_iterator == m_untouched_memory_end) {
+            // slow path, only happens when a new block needs to be allocated
+            AllocateNewBlock();
+        }
+
+        // peel off one chunk from the untouched memory
+        auto tmp = m_untouched_memory_iterator;
```

Comment:
Perhaps add a comment here explaining that the `next` pointer of in-use elements actually doesn't matter until it's deallocated, so it doesn't need initialization here.